cpp/src/arrow/pretty_print.cc

```diff
@@ -339,6 +339,17 @@ class ArrayPrinter : public PrettyPrinter {
     children.reserve(array.num_fields());
     for (int i = 0; i < array.num_fields(); ++i) {
       children.emplace_back(array.field(i));
+      // set field array null at index which parent struct is null,
+      // because StructBuilder#AppendNull dosen't set child null bitmap.
+      auto field_null_buf = array.field(i)->null_bitmap();
+      if (field_null_buf != nullptr) {
+        uint8_t* field_null_data = field_null_buf->mutable_data();
```

Comment:
Why exactly are you mutating a null bitmap in a pretty printer?

Reply:
Thanks for review, I have some questions, correct me if I am wrong:
1 now seems AppendNull/AppendNulls set null bitmap and append empty value using data_builder_, if AppendEmpty/AppendEmpties also do this, what is the difference between this two API?
2 From http://arrow.apache.org/docs/format/Columnar.html#struct-layout that there is no need for the child arrays to be null when the top-level struct slot is null, so I thought only 'empty' value should be appended (without setting null bitmap)。So in PrettyPrint, we need to change child null bitmap to ensure the correct output.

Reply:
The difference should be this:
* `AppendNull` appends a null entry in the struct
* `AppendEmptyValue` appends a non-null entry in the struct

Reply:
> So in PrettyPrint, we need to change child null bitmap to ensure the correct output

We certainly must not *change* the null bitmap. PrettyPrint just shows the child arrays as they are, it does not need to be changed here (such arrays can already be produced otherwise).

Reply:
> The difference should be this:
> 
> * `AppendNull` appends a null entry in the struct
> * `AppendEmptyValue` appends a non-null entry in the struct

Thanks for your clarification, agree with you and I have updated this PR, please take a look, thanks!

Reply:
> > So in PrettyPrint, we need to change child null bitmap to ensure the correct output
> 
> We certainly must not _change_ the null bitmap. PrettyPrint just shows the child arrays as they are, it does not need to be changed here (such arrays can already be produced otherwise).

Ok, reverted.